bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewShadeHandler.java

```diff
@@ -596,4 +611,100 @@ private void doRefreshShade(RefreshKind kind) {
             }
         }
     }
+
+    /**
+     * Initialize the dynamic channels if the respective device supports them.
+     */
+    private void updateDynamicChannels() {
+        if (!dynamicChannelsInitialized) {
+            dynamicChannelsInitialized = dynamicChannelsInitializer();
+        }
+    }
+
+    /**
+     * Helper to do the actual work of initializing the dynamic channels if the respective device supports them.
+     *
+     * @return true if all channels were successfully initialised.
+     */
+    private boolean dynamicChannelsInitializer() {
+        final Capabilities capabilities = this.capabilities;
+        if (capabilities == null) {
+            // stuff not initialised
+            return false;
+        }
+
+        // note: this is an immutable list
+        List<Channel> channels = thing.getChannels();
+        final Stream<Channel> channelStream = channels.stream();
+
+        // predicate to filter the vane channel
+        final Predicate<Channel> vaneChannelFilter = c -> HDPowerViewBindingConstants.CHANNEL_TYPE_VANE
+                .equals(c.getChannelTypeUID());
+
+        // predicate to filter the secondary channel
+        final Predicate<Channel> secChannelFilter = c -> HDPowerViewBindingConstants.CHANNEL_TYPE_SECONDARY
+                .equals(c.getChannelTypeUID());
+
+        // current and required state of the vane channel
+        final boolean vaneChannelExisting = channelStream.anyMatch(vaneChannelFilter);
+        final boolean vaneChannelRequired = capabilities.usesVaneChannel();
+
+        final boolean secChannelExisting = channelStream.anyMatch(secChannelFilter);
+        final boolean secChannelRequired = capabilities.usesSecondaryChannel();
+
+        if ((vaneChannelExisting == vaneChannelRequired) && (secChannelExisting == secChannelRequired)) {
+            // nothing to do
+            return true;
+        }
+
+        // make a mutable copy of the original immutable channel list
+        channels = new ArrayList<>(channels);
+
+        boolean error = false;
+        if (vaneChannelRequired) {
+            // build the vane channel
+            Channel vaneChannel = ChannelBuilder
+                    .create(new ChannelUID(thing.getUID(), HDPowerViewBindingConstants.CHANNEL_SHADE_VANE))
+                    .withType(HDPowerViewBindingConstants.CHANNEL_TYPE_VANE).withKind(ChannelKind.STATE)
+                    .withAcceptedItemType(CoreItemFactory.DIMMER)
+                    .withDescription(translater.getText("channel-type.hdpowerview.shade-vane.description"))
+                    .withLabel(translater.getText("channel-type.hdpowerview.shade-vane.label")).build();
+
+            // add the vane channel
+            error |= !channels.add(vaneChannel);
```

Comment:
If `error` is `true` after this, why continue (everything will be dropped in line 699 anyway)?

Reply:
Makes sense.